pennylane/capture/meta_type.py

```diff
@@ -0,0 +1,129 @@
+    # pennylane/capture/meta_type.py:37:4: W0223: Method 'update' is abstract in class 'AbstractValue' but is not overridden in child class 'AbstractOperator' (abstract-method)
+
+    class AbstractOperator(jax.core.AbstractValue):
+        """An operator captured into plxpr."""
+
+        def __eq__(self, other):
+            return isinstance(other, AbstractOperator)
+
+        def __hash__(self):
+            return hash("AbstractOperator")
+
+        def _matmul(self, *args):
+            return qml.prod(*args)
+
+        def _mul(self, a, b):
+            return qml.s_prod(b, a)
+
+        def _rmul(self, a, b):
+            return qml.s_prod(b, a)
+
+        def _add(self, a, b):
+            return qml.sum(a, b)
+
+    jax.core.raise_to_shaped_mappings[AbstractOperator] = lambda aval, _: aval
+
+    return AbstractOperator
+
+
+class PLXPRMeta(type):
+    """A metatype that:
+
+    * automatically registers a jax primitive to ``cls._primitive``
+    * Dispatches class creation to ``cls._primitive.bind`` instead of normal class
+    creation when the primitive is defined and plxpr capture is enabled.
+
+    See ``pennylane/capture/explanations.md`` for more detailed information on how this technically
+    works.
+
+    """
+
+    def __init__(cls, *_, **__):
+
+        # Called when constructing a new type that has this metaclass.
+        # Similar to __init_subclass__ , this allows us to run this code
+        # every time we define a new class
+
+        if not has_jax:
+            cls._primitive = None
+            return
+
+        cls._primitive = jax.core.Primitive(cls.__name__)
+
+        @cls._primitive.def_impl
+        def default_call(*args, **kwargs):
+            if "n_wires" not in kwargs:
+                return type.__call__(cls, *args, **kwargs)
+            n_wires = kwargs.pop("n_wires")
+            wires = args[-n_wires:]
+            args = args[:-n_wires]
+            return type.__call__(cls, *args, wires=wires, **kwargs)
+
+        # logic here will be extended when we make more things use this meta class
+        abstract_type = _get_abstract_operator()
+
+        @cls._primitive.def_abstract_eval
+        def abstract_init(*_, **__):
+            return abstract_type()
+
+    def __call__(cls, *args, **kwargs):
+        # this method is called everytime we want to create an instance of the class.
+        # default behavior uses __new__ then __init__
+        # when tracing is enabled, we want to
+
+        if not plxpr_enabled() or cls._primitive is None:
+            return type.__call__(cls, *args, **kwargs)
+        # use bind to construct the class if we want class construction to add it to the jaxpr
+        if getattr(cls, "_meta_coerce_wires", False):
+            if "wires" in kwargs:
+                wires = kwargs.pop("wires")
+                wires = tuple(wires) if isinstance(wires, Iterable) else (wires,)
+                kwargs["n_wires"] = len(wires)
+                args += wires
```

Comment:
So this always unpacks wires, right? If we have templates acting on large qubit registers (thinking ISQ circuits), will this lead to problems because of a large number of arguments?

Reply:
So yes, ongoing discussion.  I'm actually thinking of adding a toggle between the two different ways of handling wires.

Reply:
Is there any downside to packing the wires into a single array?

Reply:
I guess we normally pack them into a `Wires` object, and wires isn't jax friendly.  We could potentially examine making `Wires` work better with program capture as well.

Reply:
Well here each wire still has to be jax friendly, even if the `Wires` object is not, right?
Or do you want to enable that each wire can have a different type as long as that type is jax friendly?

I think this flexibility might not be that useful, in the end the wires have to be integers to be used effectively as qubit indices when interfacing with low-level devices and the compiler.